src/toil/jobStores/abstractJobStore.py

```diff
@@ -1958,3 +1986,142 @@ def _get_is_directory(cls, url: ParseResult) -> bool:
     def _list_url(cls, url: ParseResult) -> list[str]:
         # TODO: Implement HTTP index parsing and FTP directory listing
         raise NotImplementedError("HTTP and FTP URLs cannot yet be listed")
+
+
+class FtpFsAccess:
+    """
+    FTP access with upload.
+
+    Taken and modified from https://github.com/ohsu-comp-bio/cwl-tes/blob/03f0096f9fae8acd527687d3460a726e09190c3a/cwl_tes/ftp.py#L8
```

Comment:
The license at cwl-tes is also Apache 2.0. Should I add a section to the Toil license that specifically states all code under this class is under cwl-tes's Apache 2.0 license?

Reply:
You don't need to repeat the license.

If there are code quotes, then you could mention the copyright of the original authors / copyright holders:

Copyright 2018 EMBL-EBI (Michael R. Crusoe's original implementation, written under contract)
Copyright 2019 adamstruck <adamjstruck@gmail.com>
Copyright 2020 Joris Vankerschaver <jvankerschaver@enthought.com>
Copyright 2020 Dab Brill <dbrill@enthought.com>
Copyright 2020 Manabu ISHII <manabu.ishii.rb@gmail.com>

Otherwise if this implementation is different enough, you don't need to mention the authors.

I suggest this metric: if we wanted to change the license of Toil, would you have to get permission from these copyright holders because the code is essentially the same? Or is the code so adapted or different for other reasons that it is a new work?

Reply:
One approach is just add their Apache 2.0 license's copyright line to Toil's LICENSE file along with the others.

You could also drop it in a different file and import it to here, and in that file have `Copyright 2017 Oregon Health and Science University` and the copyright boilerplate we use for our files. I think.

Also the `#8` at the end of this link is pointing to the wrong line number.

Reply:
I think it's easier to do the latter by putting the code in a separate file, so I'll do that